lib/geoengineer/gps/gps.rb

```diff
@@ -194,67 +194,71 @@ def loop_projects_hash(projects_hash)
           nodes.each_pair do |node_type, node_names|
             node_names.each_pair do |node_name, attributes|
               node_type_class = find_node_class(node_type)
-              node = node_type_class.new(project, environment, configuration, node_name, attributes)
-              yield environments, configurations, nodes, node_names, node
+              yield node_type_class.new(project, environment, configuration, node_name, attributes)
             end
           end
         end
       end
     end
   end
 
-  def expand_meta_nodes(projects_hash)
-    # Bit of a hack just execute expand thre times to support meta-meta-meta nodes
-    ret, = expand_meta_node_layer(
-      *expand_meta_node_layer(
-        *expand_meta_node_layer(projects_hash, [])
-      )
-    )
-    ret
+  def expand_meta_node(node)
+    children_nodes = GeoEngineer::GPS.deep_dup(node.build_nodes)
+    children_nodes.reduce(children_nodes.clone) do |expanded, (node_type, node_names)|
+      node_names.reduce(expanded.clone) do |inner_expanded, (node_name, attributes)|
+        node_type_class = find_node_class(node_type)
+        node = node_type_class.new(node.project, node.environment, node.configuration, node_name, attributes)
+        next inner_expanded unless node.meta?
+
+        node.validate # ensures that the meta node has expanded and has correct attributes
+        deep_merge(inner_expanded, expand_meta_node(node))
+      end
+    end
   end
 
-  def expand_meta_node_layer(projects_hash, previously_built_nodes)
-    all_built_nodes = []
+  def expand_meta_nodes(projects_hash)
     # We dup the original hash because we cannot edit and loop over it at the same time
-    loop_projects_hash(GeoEngineer::GPS.deep_dup(projects_hash)) do |_, _, _, _, node|
+    loop_projects_hash(GeoEngineer::GPS.deep_dup(projects_hash)) do |node|
       next unless node.meta?
       node.validate # ensures that the meta node has expanded and has correct attributes
```

Comment:
move this into the `expand_meta_node` method rather than always calling before expand